NumericField: fix a number of issues, i.e. typing numbers in French and other locales

mckaragoz opened the pull request:
We handled and should be solved (especially on server-side) several issues: #2146, #2487, #2583, #2609, #2632; and it complete and works well with #2626.

1. MudNumericField now didn't accept letter keys on pressed (before it allows, but removes when lost focus)
2. Added AZERTY support.
3. Before when we press up or down button on `Immediate="False"` numeric field, it increase or decrease only once, and then it lost focus and pressing key will scroll the page, now it seems to accept continuously.
4. Before when we press up or down button on `Immediate="True"` numeric field, it increases or decreases, but it didn't see the last value until lost focus, now it both accept and show value.
5. Fix Enter and NumpadEnter buttons that they can eligible for users' own keypress events.

UPDATE:
6. Added support for Select All, Copy and Paste (Ctrl+a, Ctrl+c, Ctrl+v)

Before:

https://user-images.githubusercontent.com/78308169/132111218-7f61a01b-e43a-45b2-8231-969ea21f555e.mp4

After:

https://user-images.githubusercontent.com/78308169/132111219-a632eaf7-63ae-4e1d-92ba-9c2e92b629fc.mp4

fixes #2146, fixes #2487, fixes #2583, fixes #2609, fixes #2632

henon commented on src/MudBlazor.Docs/Pages/Components/NumericField/NumericFieldPage.razor:
```diff
@@ -68,7 +68,7 @@
                     You can also set the <CodeInline>DebounceInterval</CodeInline> parameter to the number of milliseconds you want to wait before updating the bound value. If you need to know when the interval elapses, you can pass an <CodeInline>OnDebounceIntervalElapsed</CodeInline> EventCallback.
                     Notice how in this example the debounced text only updates 500 ms after you stop typing.
-                    <MudAlert Severity="Severity.Warning" Dense="true" Class="mt-3">Warning: when <CodeInline>Immediate="false"</CodeInline> (default), if the user types a value and then presses the arrow up/down on the keyboard, the step is calculated on the last accepted value.</MudAlert>
+                    <MudAlert Severity="Severity.Warning" Dense="true" Class="mt-3">Warning: Pressing arrow up/down or Shift + MouseWheel do not affected by <CodeInline>DebounceInterval</CodeInline></MudAlert>
```
Warning: Pressing arrow up/down or Shift + MouseWheel are not debounced.

mckaragoz replied:
I changed this, thanks :)

henon commented on src/MudBlazor/Components/NumericField/MudNumericField.razor.cs:
```diff
@@ -144,8 +144,6 @@ protected override Task SetValueAsync(T value, bool updateText = true)
         protected override void OnBlurred(FocusEventArgs obj)
         {
             base.OnBlurred(obj);
-            if (RuntimeLocation.IsServerSide)
-                _key++; // this forces a re-render on the inner input to display the value correctly
```
Are you 100% sure this is not needed any more? I added that for a reason

mckaragoz replied:
I think its one of the reason to make 2x step when press arrow buttons. #1543. I encounter it again when i change code to prevent keypress, so i simplified the other voids like that and onkeyintercept one. I remove this but added also each place where needed, so now it can prevent the keypress correctly and can also key renew.

henon replied:
the reason for this was to re-format the input value. for instance, if you enter 1234,56 it will format to 1.234,56

mckaragoz replied:
No problem, formatting works the same.

https://user-images.githubusercontent.com/78308169/132125395-6a1be25a-1a8e-4087-a423-18498ace514e.mp4

henon commented:
@JohanArleth would you please be so kind as to check if the NumericField still behaves correctly with respect to the things we worked on in our last collab #2136?

henon commented:
@mckaragoz did you test on both BSS and WASM?

mckaragoz commented:
@henon i added WASM support. WASM have the 3rd and 4th problems as BSS have before, and I noticed we can do same process with BSS and WASM to solve this.

Only problem after this changes, when we press arrow keys, input value changed, but page is also scrolled. To solve this i added an extra preventDefault things which are have differencies between BSS and WASM and explain it comments in code.

I noticed that MudNumericField didn't accept the Ctrl+a, c, v (select all, copy, and paste) progress with keyboard, and i added support also for this.

After this all, the only little thing is, when we hold down the arrow key, page may also scroll. But user always can prevent this by pressing Shift+Arrow, so i think this is not a big problem considering current gains.

henon commented:
Awesome @mckaragoz. I would like for @JohanArleth to look over it just in case, so we don't have any accidental regressions.

JohanArleth commented:
Using the test viewer, I'm getting exceptions in the console on every arrow up/down press. It still seems to work correctly though. (remove Position="Position.Absolute" from MudAppBar to get the test viewer working. Maybe push this fix in the pr as well?)

There is inconsistent behavior when deleting the value in the field (both immediate and not). Using a nullable backing property, as in the numeric culture test component, it will first default to 0.00 when clearing the field, and on second attempt the value is still 0.00, but the field becomes empty. I'm not sure if this is a new or old behavior.

![https://i.imgur.com/U3GbTtS.gif](https://i.imgur.com/U3GbTtS.gif)

Error on arrow
![https://i.imgur.com/2Qpt8av.png](https://i.imgur.com/2Qpt8av.png)

If you feel like it, it would be great to also get support for "Clearable" on numeric fields.

mckaragoz commented:
> Using the test viewer, I'm getting exceptions in the console on every arrow up/down press.

I have found an issue maybe it related to this https://github.com/dotnet/aspnetcore/issues/26838. This exception may be occurring due to onfocusout. They moved this issue a in .Net 6 preview 4 but i don't know when they implement a solution. I think this isn't directly about us.

> remove Position="Position.Absolute" from MudAppBar to get the test viewer working. Maybe push this fix in the pr as well?

@henon i can handle it if the team want.

> I'm not sure if this is a new or old behavior.

I tried this another branch that doesn't have the commit, but the behaviour the same in test. I don't understand the reason of the behaviour, maybe we should add a new issue after implement this. I add clear button to the test which set value to null on click, and it works properly, so it may about the backspace key.

> If you feel like it, it would be great to also get support for "Clearable" on numeric fields.

Clearable icon is overlapping the HideSpinButton, and im not good at CSS :) we need to change their position to opposite of other or when we opened one, the other automatically hide.

henon commented:
OK, sure, you can PR the testviewer fix as well. Thanks. 
I will merge this, as it seems to work fine. Clearable support can be done in a new PR

henon commented:
Big thanks to @mckaragoz ! You saved me a lot of headache by fixing this.
